_P127_Teleinfo.ino

```diff
@@ -47,250 +68,132 @@ boolean Plugin_127(byte function, struct EventStruct *event, String& string)
-        P127_URL = String(deviceTemplate[1]);
-        Serial.begin(1200, SERIAL_7E1); //Liaison série avec les paramètres
+        // Init Serial
+        const int16_t serial_rx = CONFIG_PIN1;
+        const int16_t serial_tx = CONFIG_PIN2;        
+        P127_easySerial = new ESPeasySerial(serial_rx, serial_tx);
+
+        String log = F("P127 : Init ");
+        if (PCONFIG(0)){
+          // Init teleinfo and serial for HISTORIQUE mode
+          tinfo.init(true);          
+          P127_easySerial->begin((unsigned long)1200, SERIAL_7E1, SERIAL_RX_ONLY);
+          log += F("HISTORIQUE mode 1200_7E1 RX=");          
+        }else{
+          // Init teleinfo and serial for STANDARD mode
+          tinfo.init(false);          
+          P127_easySerial->begin((unsigned long)9600, SERIAL_7E1, SERIAL_RX_ONLY);
+          log += F("STANDARD mode 9600_7E1 RX=");          
+        }
+        
+        log += serial_rx;        
+        log += F(" reading ");       
+        log += ExtraTaskSettings.TaskDeviceValueNames[0];
+        log += F(",");       
+        log += ExtraTaskSettings.TaskDeviceValueNames[1];
+        addLog(LOG_LEVEL_INFO, log);
+        
         Plugin_127_init = true;
         success = true;
         break;
       }
 
-
-    case PLUGIN_SERIAL_IN:
+    case PLUGIN_READ:
       {
-        uint8_t serial_buf[P127_BUFFER_SIZE];
-        int RXWait = 10;
-        int timeOut = RXWait;
-        size_t bytes_read = 0;
-        char log[40];
-
+        if ( ! Plugin_127_init ) break ;
 
-        while (timeOut > 0)
-        {
-          while (Serial.available()) {
-            if (bytes_read < P127_BUFFER_SIZE) {
-              serial_buf[bytes_read] = Serial.read();
-              bytes_read++;
-            }
-            else {
-              Serial.read();  // when the buffer is full, just read remaining input, but do not store...
-
-            }
-            timeOut = RXWait; // if serial received, reset timeout counter
-          }
-          delay(1);
-          timeOut--;
-        }
+        event->sensorType = SENSOR_TYPE_DUAL;
+        success = true;
+        int value;
+        char cvalue[TAILLE_MAX_VALUE]; 
+        char *ret;
 
-        if (bytes_read == P127_BUFFER_SIZE)  // if we have a full buffer, drop the last position to stuff with string end marker
+        String log = F("P127 : ");
+        
+        for (byte x = 0; x < 2; x++)
         {
-          while (Serial.available()) { // read possible remaining data to avoid sending rubbish...
-            Serial.read();
-          }
-          bytes_read--;
-          // and log buffer full situation
-          // strcpy_P(log, PSTR("Teleinfo: serial buffer full!"));
-          // addLog(LOG_LEVEL_ERROR, log);
+          log += ExtraTaskSettings.TaskDeviceValueNames[x];          
+          log += F("=");          
+          ret = tinfo.valueGet(ExtraTaskSettings.TaskDeviceValueNames[x], cvalue);                          
```

Comment:
Does the `valueGet` function make sure the content of `cvalue` is null-terminated?

Reply:
`cvalue` was copied using `strncpy` using strlen of another pointer 
https://github.com/marco402/LibTeleinfo/blob/rewrite-Wifinfo/examples/Wifinfo/LibTeleinfo.cpp#L383
so I guess that `cvalue` may not be null-terminated and is not safe.
What would be the good direction ?

Reply:
Well at first, if external files are needed to make it compile, then you should include them, or at least post a link to them in comments at the include.

Otherwise it is next to impossible for others to use the code.

[strncpy](http://www.cplusplus.com/reference/cstring/strncpy/) does not append a 0.
You can use the macro defined in ESPEasy called `ZERO_TERMINATE` to fix that, or even better, use the macro `ZERO_FILL` before calling this copy function.

Reply:
`cvalue` has been ZERO_FILLED